impl/client/src/main/javascript/web/pentaho/theme/main.js

```diff
@@ -0,0 +1,146 @@
+/*!
+ * Copyright 2018 Hitachi Vantara. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+define([
+  "require",
+  "module",
+  "pentaho/config/service",
+  "pentaho/module/util",
+  "pentaho/shim/es6-promise"
+], function(localRequire, module, configService, moduleUtil) {
+
+  "use strict";
+
+  // "$_$_1", "$_$_2", ...
+  var SELF_PREFIX = "$_$_";
+  var SELF_REGEX = /^\$_\$_/;
+  var selfCounter = 0;
+
+  /**
+   * The `pentaho/theme!` module is an AMD/RequireJS loader plugin that
+   * loads a configured theme module for another given module.
+   *
+   * **AMD Plugin Usage**: `"pentaho/theme!{moduleId}"`
+   *
+   * 1. `{moduleId}` — The identifier of the module being themed. To refer to the requesting module,
+   *    use the special value `_`.
+   *
+   * To register a theme module, configure the module composed of the plugin and its argument.
+   * For example, to configure the theme of the `my/View` module to be the `my/theme/view` module,
+   * specify the following AMD/RequireJS configuration:
+   *
+   * ```js
+   * {
+   *   "config": {
+   *     "pentaho/theme!my/View": {
+   *       main: "my/theme/view"
```

Comment:
All these moduleIDs must be versioned by a requirejs manager plugin.

Reply:
We've talked in the meantime. The conclusion is that the `config` keys are already being versioned. As for the the values, a custom plugin needs to exist that knows the structure of the actual configuration. We'd move on without such a plugin for now, as users can still use configuration file rules to achieve the same effect, and in that context a means will exist to version these ids (by using `module.resolveId()`).